contactaccessrights/js/src/access-rights/controllers/access-rights-modal-ctrl.js

```diff
@@ -0,0 +1,116 @@
+define([
+  'common/lodash',
+  'access-rights/modules/controllers'
+], function (_, controllers) {
+  'use strict';
+
+  controllers.controller('AccessRightsModalCtrl', [
+    '$q', '$modalInstance', 'Region', 'Location', 'Right',
+    function ($q, $modalInstance, Region, Location, Right) {
+      var vm = this;
+
+      vm.availableData = {
+        regions: [],
+        locations: []
+      };
+
+      vm.selectedData = {
+        locations: [],
+        regions: []
+      };
+
+      vm.originalData = {
+        locations: [],
+        regions: []
+      };
+
+      vm.errorMsg = '';
+
+      vm.dataLoaded = false;
+
+      /**
+       * Closes the modal
+       */
+      vm.cancel = function () {
+        $modalInstance.dismiss('cancel');
+      };
+
+      /**
+       * Saves data and closes the modal
+       */
+      vm.submit = function () {
+        $q.all([persistValues('regions'), persistValues('locations')])
+          .then(function () {
+            $modalInstance.dismiss('cancel');
+          })
+          .catch(function (err) {
+            vm.errorMsg = 'Error while saving data';
+          });
+      };
+
+      /**
+       * Saves the new values, and deletes the removed ones
+       *
+       * @param  {string} type  Either "regions" or "locations"
+       * @return {Promise}      The result of all promises
+       */
+      function persistValues(type) {
+        var originalData = vm.originalData[type];
+        var selectedData = vm.selectedData[type];
+
+        var originalEntityIds = originalData.map(function (i) {
+          return i.entity_id;
+        });
+        var newEntityIds = _.difference(selectedData, originalEntityIds);
+        var removedRightIds = _.difference(originalEntityIds, selectedData)
+          .map(function (entityId) {
+            return _.find(originalData, function (i) {
+                return i.entity_id === entityId;
+              }).id;
+          });
+
+        var promises = [];
+
+        if (newEntityIds.length > 0) {
+          promises.push(Right['save' + _.capitalize(type)](newEntityIds));
+        }
+        if (removedRightIds.length > 0) {
+          promises.push(Right.deleteByIds(removedRightIds));
+        }
+
+        return $q.all(promises);
+      }
+
+      /**
+       * Load the API data
+       */
+      $q.all([Region.getAll()
```

Comment:
There is a lot of repetition here, you are basically doing the same thing with both promises' results.
What about something like (not sure it's the optimal way, but you get the idea)
```js
$q.all([
  Region.getAll(),
  Location.getAll()
])
.then(function (values) {
  return { regions: values[0], locations: values[1] };
})
.then(function (values) {
  return _.map(values, function (value, key) {
    vm.availableData[key] = value.values;
    return Right['get' + _.capitalize(key)];
  });
})
.then(function (values) {
  return { regions: values[0], locations: values[1] };
})
.then(function (values) {
  _.forEach(function (value, key) {
    vm.originalData[key] = value.values;
    vm.selectedData[key] = value.values.map(function (entity) {
      return entity.entity_id;
    });
  });
});
```